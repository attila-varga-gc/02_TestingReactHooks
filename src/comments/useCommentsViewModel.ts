import { useState, useCallback, useEffect } from "react";
import { CommentsModel } from "./useCommentsModel";

export type CommentViewModel = {
  id: number;
  message: string;
  likes: string;
};

export type CommentsViewModelState = {
  comments: CommentViewModel[];
};

export type CommentsViewModel = CommentsViewModelState & {
  like: (comment: CommentViewModel) => Promise<void>;
};

export function useCommentsViewModel({
  comments,
  addLike,
}: CommentsModel): CommentsViewModel {
  const [viewModel, setViewModel] = useState<CommentsViewModelState>({
    comments: [],
  });
  useEffect(
    () =>
      setViewModel((viewModel) => ({
        ...viewModel,
        comments: comments.map(({ likes, ...comment }) => ({
          ...comment,
          likes: `${likes} likes`,
        })),
      })),
    [comments]
  );
  const like = useCallback(
    (commentViewModel: CommentViewModel) => {
      return addLike(commentViewModel.id);
    },
    [addLike]
  );
  return {
    ...viewModel,
    like,
  };
}
