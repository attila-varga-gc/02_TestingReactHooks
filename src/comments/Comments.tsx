import { CommentsViewModel } from "./useCommentsViewModel";

export function Comments({ comments, like }: CommentsViewModel) {
  return (
    <div>
      {comments.map((comment) => (
        <div key={comment.id}>
          <div>{comment.message}</div>
          <div>{comment.likes}</div>
          <button type="button" onClick={() => like(comment)}>
            Like
          </button>
        </div>
      ))}
    </div>
  );
}
