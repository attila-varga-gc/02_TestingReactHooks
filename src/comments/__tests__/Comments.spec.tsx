import * as Renderer from "react-test-renderer";
import { Comments } from "../Comments";
import { CommentsViewModel } from "../useCommentsViewModel";

it("should test against the snapshot", () => {
  const model = Mocked<CommentsViewModel>({
    comments: [
      {
        id: 42,
        message: "Test",
        likes: "12 likes",
      },
    ],
    like: jest.fn(),
  });
  expect(Renderer.create(<Comments {...model} />).toJSON()).toMatchSnapshot();
});

it("should test the click event", () => {
  const model = Mocked<CommentsViewModel>({
    comments: [{ id: 1 }],
    like: jest.fn(),
  });
  const tree = Renderer.create(<Comments {...model} />);
  tree.root.findByType("button").props.onClick();
  expect(model.like).toHaveBeenCalledTimes(1);
  expect(model.like).lastCalledWith(model.comments[0]);
});
