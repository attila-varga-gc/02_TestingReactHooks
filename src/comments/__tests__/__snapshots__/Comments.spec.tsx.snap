// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test against the snapshot 1`] = `
<div>
  <div>
    <div>
      Test
    </div>
    <div>
      12 likes
    </div>
    <button
      onClick={[Function]}
      type="button"
    >
      Like
    </button>
  </div>
</div>
`;
